### PR TITLE
ARROW-10008: [C++][Dataset] Fix filtering/row group statistics of dict columns

### PR DESCRIPTION
**Pull request by bkietz**
Parquet row group statistics did not respect dict encoding. Also added a workaround to support filtering a dictionary encoded column.

### CODE REVIEW

File: cpp/src/arrow/dataset/filter.cc [hunk 1]
```diff
@@ -1255,14 +1255,42 @@ struct TreeEvaluator::Impl {
...
-    ARROW_ASSIGN_OR_RAISE(auto rhs, Evaluate(*expr.right_operand()));
+    ARROW_ASSIGN_OR_RAISE(Datum lhs, Evaluate(*expr.left_operand()));
+    ARROW_ASSIGN_OR_RAISE(Datum rhs, Evaluate(*expr.right_operand()));
 
     if (IsNullDatum(lhs) || IsNullDatum(rhs)) {
       return Datum(std::make_shared<BooleanScalar>());
     }
 
-    DCHECK(lhs.is_array());
+    if (lhs.type()->id() == Type::DICTIONARY && rhs.type()->id() == Type::DICTIONARY) {
```
Comment (bkietz) [inline cpp/src/arrow/dataset/filter.cc:1265]: @wesm What do you think about adding kernels to scalar_compare.cc which do this inside compute:: ?

File: cpp/src/arrow/dataset/file_parquet.cc [hunk 2]
```diff
@@ -171,6 +170,15 @@ static std::shared_ptr<StructScalar> ColumnChunkStatisticsAsStructScalar(
     return nullptr;
   }
 
+  auto maybe_min = min->CastTo(field->type());
+  auto maybe_max = max->CastTo(field->type());
```
Comment (jorisvandenbossche) [inline cpp/src/arrow/dataset/file_parquet.cc:174]: Does this change behaviour? For a dictionary with string values, is field->type() string or dictionary?

**Review by jorisvandenbossche — COMMENTED** [review]
For me the non-performant way of decoding is fine for now (certainly because the array+scalar case will be more common). 

But should there be some more tests added?

Could also use the small reproducer from the issue (my comment) to add as a python test

File: cpp/src/arrow/dataset/file_parquet.cc (hunk 2 above)
Comment (bkietz) [reply cpp/src/arrow/dataset/file_parquet.cc:174]: `StatisticsAsScalars` returns scalars whose types are the correct *physical type*, so even if the column was `dictionary(string)` min and max would be just `string` before this cast

Comment (bkietz) [reply cpp/src/arrow/dataset/file_parquet.cc:174]: (IE, it only changes behavior in cases where the physical type wasn't appropriate)

**Review by jorisvandenbossche — APPROVED** [review]

File: cpp/src/arrow/dataset/filter.cc (hunk 1 above)
Comment (wesm) [reply cpp/src/arrow/dataset/filter.cc:1265]: Yes, this sounds fine, can you open a JIRA issue about it?

**Pull request closed without merging**